File: matrixDiagonalOrder.py
# Time Complexity: O(mn), where m and n are dimensions of matrix as we go over each matrix element once
# Space Complexity: O(1), We only create a result array of dimensions mxn to store and return the result
def findDiagonalOrder(mat: list[list[int]]) -> list[int]:
        if not mat or len(mat) == 0:
            return []
        m = len(mat)
        n = len(mat[0])
        result = [0]*(m*n)
        i,r,c = 0,0,0
        dir = 1
        # dir = 1 bot to top
        # dir = -1 top to bot
        while i < m*n:
            result[i] = mat[r][c]
            if dir == 1:
                if c == n-1: # We hit the end of a column, only increment the row and flip directions
                    dir = -1
                    r+=1
                elif r == 0: # We are at the beginning row, only increment columns, and flip dirs
                    dir = -1
                    c+=1
                else:   # increment cols and decrement row otherwise
                    r-=1
                    c+=1
            else:
                if r == m-1:  # We are at the end of rows, increment only cols and flip dirs
                    dir = 1
                    c+=1
                elif c == 0:  # We are at the beginning col, only increment rows, and flip dirs
                    dir = 1
                    r+=1
                else:   # increment rows and decrement cols otherwise
                    r+=1
                    c-=1
            i+=1
        return result

File: MatrixDiagonalOrder.java
//TC O(m*n)
//SC O(1)
//works in leetcode
class MatrixDiagonalOrder {

    private static final int  UP=0, DOWN=1;

    public static int[] findDiagonalOrder(int[][] mat) {

        int m= mat.length;
        int n=mat[0].length;
        int [] result = new int [m*n];
        int resultIndex =0;

        int direction = UP;
        int row=0,col=0;

        while(resultIndex != m*n){
            result[resultIndex] = mat[row][col];
            resultIndex++;

            if(direction == UP){
                if(col==n-1){
                    direction = DOWN;
                    row++;
                }
                else if(row == 0){
                    direction = DOWN;
                    col++;
                }
                else{
                    row--;
                    col++;
                }

            }else
            {

                if(row == m-1){
                    direction = UP;
                    col++;
                }
                else if(col ==0){
                    direction = UP;
                    row++;
                }else{
                    row++;
                    col--;
                }

            }

        }

        return result;

    }


    public static void main(String [] args){
        int [][] matrix ={{1,2,3},{4,5,6},{7,8,9}};

        int [] out = findDiagonalOrder(matrix);

        for(int i=0; i < out.length;i++)
            System.out.println(out[i]);


    }
}